wallet_migrator/src/lib.rs

```diff
@@ -0,0 +1,65 @@
+pub mod credx2anoncreds;
+pub mod error;
+pub mod vdrtools2credx;
+
+use std::fmt::Display;
+
+use error::MigrationResult;
+use log::{debug, error, info, warn};
+pub use vdrtools::{
+    types::domain::wallet::{Config, Credentials, KeyDerivationMethod, Record},
+    Locator, WalletHandle,
+};
+
+/// Retrieves all records from a wallet and migrates them
+/// by applying the `migrate_fn` argument.
+///
+/// The migrated records are inserted into a newly created
+/// wallet, based on the provided `config` and `credentials`.
+pub async fn migrate_wallet<E>(
+    wallet_handle: WalletHandle,
+    config: Config,
+    credentials: Credentials,
+    migrate_fn: impl FnMut(Record) -> Result<Option<Record>, E>,
+) -> MigrationResult<()>
+where
+    E: Display,
+{
+    info!("Starting wallet migration...");
+    let locator = Locator::instance();
+
+    debug!("Creating new wallet {} ...", config.id);
+    locator
+        .wallet_controller
+        .create(config.clone(), credentials.clone())
+        .await?;
+
+    debug!("Opening new wallet...");
+
+    let new_wh = locator
+        .wallet_controller
+        .open(config.clone(), credentials.clone())
+        .await?;
+
+    debug!("Migrating records from wallet with handle {wallet_handle:?} to wallet with handle {new_wh:?}");
+
+    let res = locator
+        .wallet_controller
+        .migrate_records(wallet_handle, new_wh, migrate_fn)
+        .await;
+
+    locator.wallet_controller.close(new_wh).await?;
```

Comment:
This can potentially fail and then the following wallet deletion won't be run. 

I would rather:
- have dry run option I can use first
- on failure not delete at all (to avoid this issue with skipping deletion if closing fails - just make it users responsibility to manage wallets they might have created while using this)
- make migration idempotent, so I could migrate to same target, even if I initially get failure on attempt to migrate some wallet item

Reply:
Looking at the wallet closing code, it can really fail only for two reasons: mutex poisoning or unknown wallet handle.

The latter is practically impossible since we create the wallet and the former is not really something within our control (the mutex gets poisoned if a panic happens in a different thread that acquired it but didn't get the chance to release it).

I could tweak `libvdrtools` a bit but I think the end result is the same, so instead I'll just log the error from closing the wallet, just like the one from the delete call (which can theoretically also fail).

In the event of an error, this still allows the user to be able to close/delete the wallet themselves (they provided all the stuff required for creating and opening it).